### PR TITLE
Rewrite polling

### LINKED ISSUE #735: Long Polling drops after 1 day
* Telegraf.js Version: 3.32.0
* Node.js Version: v12.5.0
* Operating System: Ubuntu 18.04

Hi, I have a problem where the bot stops responding to any message after 1 day or so. I use long polling method because my server is in a corporate intranet with proxy connection to internet.  So webhook is no option for me. 

I use PM2 and the process is still running fine, the bot is still running. no errors from log.  I've been searching for answers for 3 days and haven't found it yet, so I'm writing here. please help..

### LINKED ISSUE #851: Processing an update twice when graceful shutdowned
## Context

Hey,

I've implemented graceful shutdown for my bot and face an issue. I've created a small demo project: https://github.com/mbelsky/tf-update-issue/blob/master/index.js

* Telegraf.js Version: 3.35.0
* Node.js Version: 12.13.0
* Operating System: MacOS 10.15.1

## Expected Behavior

An updated handled once

## Current Behavior

An updated handled twice

### Steps to Reproduce

1. `npm start`
2. Send '/long' message to a bot
3. Send SIGINT to the node process while it processing update
4. Let the process die
5. `npm start` again 

Screenshot 2020-01-03 at 01 29 35

As you may see it processes `933862816` twice.

I think it could be fixed with adding `offset` arg in `Telegraf. startPolling` method, but before making a PR I'm interested to know your opinion about the preferred way to fix this issue.

### LINKED ISSUE #1023: What is "Failed to fetch updates"?
I got this error, even I'm not using bot.
`Failed to fetch updates. Waiting: 1s request to https://api.telegram.org/bot......botToken......./getUpdates?offset=0&limit=100&timeout=30 failed, reason: read ECONNRESET`
**What is it? And how can I turn off it? In what kind of reason it can be?**
**UPDATED**: I got this error after some moment of sending message from another program side of api: 
```
const Telegraf = require('telegraf');
const bot_token = config.get('telegram.bot-token');
const bot = new Telegraf(bot_token);
```
And I send message like this -> 
```
bot.telegram.sendMessage(adminChannelId, orderMessage, {parse_mode: "HTML"});
return jsonPreProcessor.success({payment: payment, orderArrivals: orderArrivals});
```
Is it normal? Or I did something wrong? Telegram bot is starting from server.js with bot.launch();

### PR DESCRIPTION
**Pull request by wojpawlik**
I set out to fix #851. To keep my sanity, I extracted all polling logic into a separate class. Also fixed some other issues.

- Fixes #735, `handlerTimeout` now defaults to 5 seconds.
- Fixes #851.
- Fixes #1023, the error is now logged using `debug` and thus isn't displayed by default.
- #1104 throws a nicer error.

Also removed some barely useful configuration options.

Perhaps webhook logic should be extracted to a separate file as well? https://github.com/garronej/denoify/issues/13#issuecomment-689076929

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

## How Has This Been Tested?

By starting and stopping a bot in various ways :)

**Test Configuration**: Node.js Version: 14.0.0

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: package.json
Comment (KnorpelSenf) [inline package.json:51]: Can I assume that you verified that there is no way around adding another dep if we want to process updates reliably around starting and stopping bots?

**Review by KnorpelSenf — COMMENTED** [review]
Looks pretty good so far. I will checkout your code and play around with that tomorrow, so I can confirm that it works.

File: package.json
Comment (wojpawlik) [reply package.json:51]: Except for reimplementing it or dropping support for Node <15, there's no way.

Comment (KnorpelSenf) [reply package.json:51]: Yeah I see now that it's the way to go if you use `node-fetch` which we do.

File: src/core/network/polling.ts
Comment (KnorpelSenf) [inline src/core/network/polling.ts:40]: I can see that you copied this from the old code, but why is it even that we only rethrow upon 401 and 409? Wouldn't it make more sense to rather check for 429 or for `retry_after` and throw in all other cases?

File: src/core/network/polling.ts
Comment (KnorpelSenf) [inline src/core/network/polling.ts:17]: Fancy and I like it, but should this be `private`?

File: src/core/network/polling.ts
Comment (KnorpelSenf) [inline src/core/network/polling.ts:49]: It is not clear why we need this call. Please add a brief comment explaining why it is necessary to fetch an update and immediately discard it.

Also, it seems to be called inconsistently depending on where an error is thrown, but I cannot really pinpoint this without knowing what it actually does.

Is it relevant here to pass `allowed_updates`?

File: src/telegraf.ts
Comment (KnorpelSenf) [inline src/telegraf.ts:63]: As you're touching this file anyway, it might be the right time to rename `TContext` to `C` here, as well.

File: src/telegraf.ts
Comment (KnorpelSenf) [inline src/telegraf.ts:74]: This does not belong into `telegraf` itself, it should be the bot that determines its exit codes.

However, it certainly does belong into `bin/telegraf`, so maybe we should move this line there.

File: src/telegram.ts
Comment (KnorpelSenf) [inline src/telegram.ts:46]: Why adding `@depecated`? If a user what's to get the raw updates directly for whatever, this should be supported.

However, I do agree that we should add a comment here, explaining why this method is rarely useful. It must not become a pitfall for new users, so the method should never be called by accident.

**Review by KnorpelSenf — COMMENTED** [review]
Great work. I just have a few minor things and questions:

File: package.json
Comment (wojpawlik) [reply package.json:51]: We could alternatively use more lightweight but less popular https://www.npmjs.com/package/node-abort-controller, I'll try it tomorrow

**Comment by KnorpelSenf** [comment]
Does this introduce breaking changes or rather improve the reliability?

I'm interested in getting to the point where we can actually release 4.0, so anything that is not mission-critical may also be merged afterwards IMO. If this PR can make it before—even better. Still I think it makes sense to only add things to the 4.0 Roadmap #1202 that really need to be there (because they break stuff).

**Comment by wojpawlik** [comment]
> Does this introduce breaking changes

Yes, `LaunchOptions` are revamped, so it'll be awkward if it doesn't make it into 4.0.

File: src/core/network/polling.ts
Comment (wojpawlik) [reply src/core/network/polling.ts:49]: I could link to https://github.com/telegraf/telegraf/issues/851#issuecomment-736025893 in a comment...

Or, I could do something that sounds more appealing the more I think about it: call `confirmUpdates` in `finally` clause in `loop`. Disadvantage: call would happen even if not necessary. Advantages:

+ Bulletproof. It'll always be called, always exactly once (unless the process is killed of course).
+ There's no need to call `confirmUpdates` anywhere else, so it can be made `private`.


> Is it relevant here to pass `allowed_updates`?

Not really. We don't care what we get. And it defaults to value from previous call anyway.

Comment (KnorpelSenf) [reply src/core/network/polling.ts:49]: > I could link to https://github.com/telegraf/telegraf/issues/851#issuecomment-736025893 in a comment...

Yes, or just explain it :)

I won't be the last one to wonder about this, and a quick sentence or two is much more convenient.

> call `confirmUpdates` in `finally` clause in `loop`

That actually sounds much clearer. I think reducing the time from dozens of seconds to a few msec justifies one superfluous call every blue moon, just to be safe.

File: src/telegraf.ts
Comment (wojpawlik) [reply src/telegraf.ts:74]: This tries to emulate `warn-with-error-code` [unhandled rejections mode](https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode) when `warn` mode is used (default before Node 15). It effectively changes nothing in `throw`, `strict`, and `warn-with-error-code` modes. Without this, the process will cleanly exit with status code 0, perhaps fooling process manager into not restarting the process. Any better ideas?

Comment (KnorpelSenf) [reply src/telegraf.ts:74]:
```suggestion
    // set exit code to emulate `warn-with-error-code` behavior of
    // https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode
    // to prevent a clean exit despite an error being thrown
    process.exitCode = 1
```

File: bin/telegraf
Comment (wojpawlik) [inline bin/telegraf:107]: I'm considering adding these to all examples

File: src/core/network/polling.ts
Comment (wojpawlik) [reply src/core/network/polling.ts:49]: I'm still open to better `debug` message

File: src/telegraf.ts
Comment (wojpawlik) [inline src/telegraf.ts:76]: Perhaps this is redundant now, when `handleError` doesn't do much, and `confirmUpdates` cannot replace the error?

Comment (KnorpelSenf) [reply src/telegraf.ts:76]: Yes.
```suggestion
```

File: src/core/network/polling.ts
Comment (KnorpelSenf) [reply src/core/network/polling.ts]: I'll throw some suggestions at you. Just pick any, or none, or any combination:
- `Saving update offset`
- `Ensuring correct update offset`
- `Acknowledging updates`
- `Confirming received updates to Telegram`
- `Syncing update queue`
- `Securing update index`
I can keep going if you dislike all of them.

File: src/core/network/polling.ts
Comment (wojpawlik) [reply src/core/network/polling.ts:41]: What errors are we dealing with and how should we handle them?

- `AbortError` - `return`
- 401 Unauthorized - unrecoverable, `throw err`
- 409 Conflict - unrecoverable, `throw err`
- 429 Too Many Requests - if it ever happens, wait and retry
- 5xx - wait and retry
- `FetchError` - wait and retry, it's probably connectivity issue
- Did I forget anything?
- Unknown errors - `throw err` I guess?

Note that throwing here = crashing.

File: src/telegram.ts
Comment (wojpawlik) [reply src/telegram.ts:46]: My logic was, it's not used internally anymore, and probably shouldn't be used by users either. Users can use well-typed `Telegram::callApi` if they want to proceed anyway.

File: src/core/network/polling.ts
Comment (wojpawlik) [reply src/core/network/polling.ts:41]: I wish I could use `continue` here.

Comment (wojpawlik) [reply src/core/network/polling.ts:41]: > `getUpdates` never returns 429.
https://t.me/tdlibchat/5368

Comment (KnorpelSenf) [reply src/core/network/polling.ts:41]: > I wish I could use `continue` here.

Why is that not possible? :thinking:

Comment (KnorpelSenf) [reply src/core/network/polling.ts:41]: Regarding the errors, we should maybe just leave it like that and try to figure out what's best over time. I cannot come up with an ad hoc solution here, especially not when we don't have any official source that tells use which errors could be returned (do we?).

Comment (wojpawlik) [reply src/core/network/polling.ts]: >> I wish I could use `continue` here.
>
> Why is that not possible? :thinking: 

Bare `if (...) continue` wouldn't `await wait(...)`, I had to get more creative than that.

I decided to retry only on `FetchError` and `>= 500` error codes, people will open issues if it's too harsh, at least we know we aren't hiding any important errors.

File: package.json
Comment (wojpawlik) [reply package.json:51]: `node-abort-controller` produces this:

```
src/core/network/polling.ts:30:11 - error TS2345: Argument of type 'AbortController' is not assignable to parameter of type 'CallApiOptions'.
  The types of 'signal.onabort' are incompatible between these types.        
    Type '((this: AbortSignal, ev: Event) => any) | null' is not assignable to type '((this: AbortSignal, event: any) => void) | null | undefined'.       
      Type '(this: AbortSignal, ev: Event) => any' is not assignable to type 
'(this: AbortSignal, event: any) => void'.
        The 'this' types of each signature are incompatible.
          Type 'import("D:/Wojciech/JS/telegraf/node_modules/@types/node-fetch/externals").AbortSignal' is not assignable to type 'AbortSignal'.
            Types of property 'onabort' are incompatible.
              Type '((this: AbortSignal, event: any) => void) | null | undefined' is not assignable to type '((this: AbortSignal, ev: Event) => any) | null'.
                Type 'undefined' is not assignable to type '((this: AbortSignal, ev: Event) => any) | null'.
```

Sticking with well-typed `abort-controller`.

File: src/telegram.ts
Comment (wojpawlik) [reply src/telegram.ts]: I undeprecated it already, should I un-undeprecate? :)

Comment (KnorpelSenf) [reply src/telegram.ts]: No, I still think that it should not be deprecated, but I understood your point now.

I suggest to leave it undeprecated

**Comment by KnorpelSenf** [comment]
I tried out the changes on `echo-bot.js`. They work pretty flawlessly. There are two things I'd like to discuss:

1. When starting the bot `node echo-bot.js` and stopping it immediately again (`C-c` as fast as you can type), the process takes several seconds before it completes. In contrast to that, when the bot has been running for more than 5 seconds, it can complete immediately. Is this the intended behaviour (and why is it even happening)?
2. I started the bot and disconnected my machine from the network.
```
  telegraf:client HTTP call getUpdates { timeout: 50, offset: 954758540, allowed_updates: [] } +50s
  telegraf:client HTTP call getUpdates { timeout: 50, offset: 954758540, allowed_updates: [] } +5m
  telegraf:client HTTP call getUpdates { timeout: 50, offset: 954758540, allowed_updates: [] } +50s
```
Notice how the request is pending for five minutes, even though the timeout is set to 50 s. We certainly should not throw when the connection fails, this would be a pain on unstable connections. However, maybe it is also not desired that the bot would just hang indefinitely, until the connection is restored. As an example, changes to the network configuration would require a clean reconnect of the bot. This does happen frequently on mobile, but most users are probably not running their bot inside Termux or something. Are there ways to host a bot that could require this?

**Comment by wojpawlik** [comment]
> 1. When starting the bot `node echo-bot.js` and stopping it immediately again (`C-c` as fast as you can type), the process takes several seconds before it completes.

Can't reporoduce, but now I experienced something _worse_: capturing `SIGINT` _before_ `bot.polling` is set! As a result, bot keeps running and will now close dirtily on next `SIGING`. The real solution to that is top-level await...

I'll be going offline for a moment to test the no network scenario.

**Comment by wojpawlik** [comment]
Perhaps Telegraf should refuse `.launch()` after `.stop()`?

> Notice how the request is pending for five minutes

Can't reproduce that as well, maybe because I'm on Windows.

> As an example, changes to the network configuration would require a clean reconnect of the bot.

Why?

**Comment by KnorpelSenf** [comment]
> > Notice how the request is pending for five minutes
> 
> Can't reproduce that as well, maybe because I'm on Windows.

That might be it, I cannot figure it out. I might look at it again in a while if I have more time.

> > As an example, changes to the network configuration would require a clean reconnect of the bot.
> 
> Why?

Once you send the request and wait for its response, you cannot receive these packets from another connection. I don't know the technical details (I know who to ask if you're interested), but I assume that you can try this out by running the bot on a phone.

**Comment by wojpawlik** [comment]
Always willing to learn :)

Can it be fixed by setting `timeout: 60000` or higher on `fetch`?

**Comment by wojpawlik** [comment]
I added 2 minute timeout to every request, does it fix the pending requests issue? Don't forget to `npm run build` before trying :)

**Comment by KnorpelSenf** [comment]
> > > Notice how the request is pending for five minutes
> > 
> > 
> > Can't reproduce that as well, maybe because I'm on Windows.
> 
> That might be it, I cannot figure it out. I might look at it again in a while if I have more time.

Added `this.abortController` to the offset syncing call defeats the whole point, is that true? As this does indeed solve the “problem”, I assume that it's actually the HTTP request that takes so long. I suspect it's some sort of backend thing that TG cannot do setup and teardown with a split second. I might be mistaken.

> I added 2 minute timeout to every request, does it fix the pending requests issue? Don't forget to `npm run build` before trying :)

Your last commits do indeed solve the problem. The `handlerTimeout` of 5000 ms is respected, so a new request is sent every five seconds. I don't know how this relates to the two minutes you set.

**Comment by KnorpelSenf** [comment]
Please run `npm audit fix` now that you're updating `package-lock.json` anyway

**Comment by KnorpelSenf** [comment]
> but now I experienced something _worse_: capturing `SIGINT` _before_ `bot.polling` is set!

That's something I cannot reproduce but perhaps because I'm not on Windows. Sounds like it's easy to fix this with a flag, though.

**Comment by KnorpelSenf** [comment]
One more observation: as you know, it is not possible to have the same bot running twice. Terminated by other `getUpdates` request is the keyword here.

Usually, if instance A is running and you start instance B, then A's request will fail, and A exits. B continues to run, it effectively takes over control.

This PR attempts to do another `getUpdates` request upon shutdown. As a result, the following happens:
1. A calls `getUpdates`
2. B calls `getUpdates`
3. A's request is terminated
4. A shuts down
5. A tries to sync the offset, hence making a last `getUpdates` request before exiting
6. B's request is terminated
7. B shuts down
8. B syncs the offset before by calling `getUpdates`, and then exists
9. No instance is running

Generally, it is the bot developer's responsibility to make sure that there is only ever a single active instance, however, I feel like it would be better to aim for the old behaviour here.

As an example, if A's request is terminated due to B's request, it makes sense to skip step 5, i.e. syncing the offset.

File: src/core/network/polling.ts
Comment (KnorpelSenf) [inline src/core/network/polling.ts:52]: This makes it impossible to restart the bot after the error!

Also, arbitrarily rewriting instance members on the fly is generally really causing mindfuck bugs. The same solution can be achieved with a simple flag and not much more effort.

**Review by KnorpelSenf — CHANGES_REQUESTED** [review]

File: src/core/network/polling.ts
Comment (wojpawlik) [reply src/core/network/polling.ts:52]: `throw`ing here causes unhandled rejection (= process crash on Node 15, preventing any attempts at reusing `Polling`).

401 should probably also avoid confirming, since it's guaranteed to fail.

Comment (wojpawlik) [reply src/core/network/polling.ts:52]: `Polling` objects are meant to be recreated, not reused. To make reuse impossible, I can abort the controller in `finally`. It can't be replaced since it's `readonly`.

Comment (wojpawlik) [reply src/core/network/polling.ts:56]: I don't know how the member would need to be named so that the code is as clear as `this.confirmUpdates = noop`. It wouldn't be used for anything else.

**Comment by wojpawlik** [comment]
> Adding `this.abortController` to the offset syncing call defeats the whole point, is that true?

Yes. If you're gonna abort that request, might as well not make it in the first place.

> Your last commits do indeed solve the problem. The `handlerTimeout` of 5000 ms is respected, so a new request is sent every five seconds. I don't know how this relates to the two minutes you set.

Weird but ok.

Any other concerns before this can be approved?

File: src/core/network/polling.ts
Comment (KnorpelSenf) [inline src/core/network/polling.ts:79]: Just use this.
```suggestion
    try {
      for await (const updates of this) {
        await handleUpdates(updates)
      }
    } catch (err) {
      if (
        !(
          err instanceof TelegramError &&
          (err.code === 401 || err.code === 409)
        )
      ) {
        await this.confirmUpdates().catch(noop)
      }
    } finally {
      debug('Long polling stopped')
    }
```
Now you can get rid of
1) some error branching logic above
2) that unconventional method rewriting stuff ([this mistake was made before](https://github.com/telegraf/telegraf/blob/359f4f106dddccabfa6523c3bbd6ddf79a8b826a/extra.js#L45-L46))

Also, there's in fact no need for util flags and its perfectly readable.

File: src/core/network/polling.ts
Comment (KnorpelSenf) [reply src/core/network/polling.ts:56]: > causes unhandled rejection

does it really? It is awaited inside a loop, no?

File: src/core/network/polling.ts
Comment (wojpawlik) [reply src/core/network/polling.ts:79]: Pretty good, but slightly wrong: `confirmUpdates` is not called on successful graceful stop.

Comment (KnorpelSenf) [reply src/core/network/polling.ts:79]: > I can abort the controller in `finally`

That might still be a good idea if you still want to ensure that `Polling` objects won't be reused. Note that I skipped those lines in the suggestion above.

**Comment by KnorpelSenf** [comment]
> Any other concerns before this can be approved?

Nothing apart from the method rewriting.

Good job.

File: src/core/network/polling.ts
Comment (KnorpelSenf) [reply src/core/network/polling.ts:79]: I'm missing the rethrowing part! We have to rethrow the error in the new catch block of course. Don't use the suggestion as-is :smile:

File: src/core/network/polling.ts
Comment (wojpawlik) [inline src/core/network/polling.ts:83]: This feels a little redundant:

```
  telegraf:main Stopping bot... Reason: SIGINT +57s
  telegraf:polling Stopping long polling... +1m
  telegraf:polling Long polling stopped +6ms
```

```suggestion
```

Comment (KnorpelSenf) [reply src/core/network/polling.ts:83]: Good catch

File: src/core/network/polling.ts
Comment (wojpawlik) [reply src/core/network/polling.ts:56]: Yes, but `.loop` is not awaited.

File: src/core/network/polling.ts
Comment (KnorpelSenf) [reply src/core/network/polling.ts:79]: > Pretty good, but slightly wrong: confirmUpdates is not called on successful graceful stop.

Only saw that now, you're right, fixed it

Comment (KnorpelSenf) [reply src/core/network/polling.ts:79]: If the `catch` this is not DRY enough, you can move it to `confirmUpdates` internally.

Comment (wojpawlik) [reply src/core/network/polling.ts:79]: How? it has no access to `err`.

Comment (wojpawlik) [reply src/core/network/polling.ts:79]: And yes, your suggestion is not DRY enough, it's not clearly superior to mine.

Comment (wojpawlik) [reply src/core/network/polling.ts:79]: If you don't have anything better, why don't you approve this, and perhaps change it next time you'll be touching the file?

Comment (wojpawlik) [reply src/core/network/polling.ts:79]: Retrying logic could be extracted into a separate function and reused for `getMe` in `.launch()` and `.handleUpdate()`, or even exported to library users; then it could no longer concern itself with internals of `Polling`.

Comment (KnorpelSenf) [reply src/core/network/polling.ts:79]: > And yes, your suggestion is not DRY enough, it's not clearly superior to mine.

You do not understand what I was going for. However, given that you just merged before we could reach consensus, I'll clean up the mess in another PR.

**Comment by wojpawlik** [comment]
Perhaps this could improve our error handling: https://www.npmjs.com/package/is-retry-allowed

**Comment by KnorpelSenf** [comment]
> Perhaps this could improve our error handling: https://www.npmjs.com/package/is-retry-allowed

It is certainly a good idea to use this for error handling, but I'm not sure how strictly this is necessary. There is a myriad of small packages that do some things a little better.

As we rethrow the errors, it is probably simple for library users to leverage that package. We might add a section to the docs, though, and include a code snippet that shows how to perform retries the proper way.

**Comment by wojpawlik** [comment]
Errors thrown from `Polling` cannot be caught: https://github.com/telegraf/telegraf/blob/a221c8b99db3fd005c447a9728068e15962a7cb2/src/telegraf.ts#L118-L119

**Comment by KnorpelSenf** [comment]
Well that's bad. But the fix is a `catch` call and not a library.

**Pull request merged**
